text/001x-battery_improvements.md

```diff
@@ -0,0 +1,110 @@
+
+# Detailed Design
+
+There are three parts to the design:
+1. A more efficient status message
+2. A scalable battery voltage message.
+3. Mechanisms that allow the new messages to seamlessly coexist with the old one along with eventual deprecation of the older message.
+
+## BATTERY_STATUS_V2
+
+The message is heavily based on [BATTERY_STATUS](https://mavlink.io/en/messages/common.html#BATTERY_STATUS) but:
+- removes the cell voltage arrays: `voltages` and `voltages_ext`
+- adds `voltage`, the total cell voltage. This is a uint32_t to allow batteries with more than 65V.
+
+
+```xml
+    <message id="???" name="BATTERY_STATUS_V2">
+      <description>Battery information. Updates GCS with flight controller battery status. Smart batteries also use this message, but may additionally send SMART_BATTERY_INFO.</description>
+      <field type="uint8_t" name="id" instance="true">Battery ID</field>
+      <field type="uint8_t" name="battery_function" enum="MAV_BATTERY_FUNCTION">Function of the battery</field>
+      <field type="uint8_t" name="type" enum="MAV_BATTERY_TYPE">Type (chemistry) of the battery</field>
+      <field type="int16_t" name="temperature" units="cdegC" invalid="INT16_MAX">Temperature of the battery. INT16_MAX for unknown temperature.</field>
+      <field type="uint32_t" name="voltage" units="mV" invalid="[UINT32_MAX]">Battery voltage (total).</field>
+      <field type="int16_t" name="current" units="cA" invalid="UINT16_MAX">Battery current (through all cells/loads). Positive if discharging, negative if charging. UINT16_MAX: field not provided.</field>
+      <field type="int32_t" name="current_consumed" units="mAh" invalid="-1">Consumed charge, -1: Current consumption estimate not provided.</field>
+      <field type="int32_t" name="energy_consumed" units="hJ" invalid="-1">Consumed energy, -1: Energy consumption estimate not provided.</field>
+      <field type="int8_t" name="battery_remaining" units="%" invalid="-1">Remaining battery energy. Values: [0-100], -1: Remaining battery energy is not provided.</field>
+      <field type="uint32_t" name="time_remaining" units="s" invalid="UINT32_MAX">Remaining battery time (estimated), UINT32_MAX: Remaining battery time estimate not provided.</field>
+      <field type="uint8_t" name="charge_state" enum="MAV_BATTERY_CHARGE_STATE">State for extent of discharge, provided by autopilot for warning or external reactions</field>
+      <field type="uint8_t" name="mode" enum="MAV_BATTERY_MODE">Battery mode. Default (0) is that battery mode reporting is not supported or battery is in normal-use mode.</field>
+      <field type="uint32_t" name="fault_bitmask" display="bitmask" enum="MAV_BATTERY_FAULT">Fault/health indications. These should be set when charge_state is MAV_BATTERY_CHARGE_STATE_FAILED or MAV_BATTERY_CHARGE_STATE_UNHEALTHY (if not, fault reporting is not supported).</field>
```

Comment:
I feel like `charge_state`, `mode`, and `fault_bitmask` could be combined into a single `uint32_t status`... the concept of a "charge state" doesn't make a lot of sense. It's either charging, discharging, or idle. If it's not charging/discharging due to some error, that would be indicated using one of the other fault flags. Having separate enums to convey some "charge state" is effectively redundant, since it's just telling you to look at the fault flags to understand why the "charge state" is abnormal. 

I would argue the same about "mode". We can convey the entire operating state of the battery with a single uint32 set of bit flags.

Reply:
Seems @WickedShell shares your opinion: https://github.com/mavlink/rfcs/pull/19/files#r838110393

I think the charging state and faults could easily be combined, and that the charging state needs to be thought out a bit more. Whether or not we can do this depends on who has implemented the current message.

Reply:
>though you might argue the charge states are logically separate

How? The battery is either between 0% and 100% or it is in fault. If it's in fault and can't charge, it would be indicated by a flag. For example if it was overdischarged, a cell is in fault, or temp is too high/low

Reply:
>The one case you can't infer is MAV_BATTERY_CHARGE_STATE_CHARGING

If we reworked this to combine charge_state/mode/fault all under a single MAV_BATTERY_V2_STATUS_FLAGS and used the same flags as the DroneCAN PR I have open, we could represent all of the information we could possibly want. You could infer charging/discharging with the flags `STATUS_FLAG_IN_USE` and `STATUS_FLAG_CHARGING`

I also think many of the existing MAV_BATTERY_FAULT fields are useless. All of the "incompatible" fields don't make any sense, neither does the "voltage spikes" field. I know undoing things in mavlink can be hard, so that's why I am suggesting we might be better off using a new bitmask enum